More improvements to HDF5 interface and Python API

paulromano opened the pull request:
Here's yet another pull request that adds improvements to how we handle output. A list of changes is as follows:

- Instead of writing strings as variable-length to HDF5 files, they are written as fixed-length C-style null-terminated strings for maximum compatibility.
- A new pair of procedures read_string_1D/write_string_1D allows 1D arrays of strings to be written/read to HDF files. I take advantage of this and write tally score moment orders as an array of strings instead of having a separate dataset for each. Much cleaner!
- Some unused procedures for writing ASCII summary files (which no longer exist) were removed.
- All the UNIT_* constants were removed and the NEWUNIT argument of open() is consistently used for Fortran I/O.
- The openmc.statepoint.StatePoint class now makes judicious use of properties to grab data from the HDF5 file as needed. This eliminates all the `_read_*` methods as well as `StatePoint.read_results`, `Tally.compute_mean` and `Tally.compute_std_dev`. Tally data is read on-the-fly as it is requested.
- Related to the last point, global_tallies are now represented as a structured array with extra fields 'mean', 'std_dev', and 'name'. The `StatePoint.compute_stdev` and `StatePoint.compute_ci` methods are gone. This does imply that there is no automated method for calculating confidence intervals anymore. Should be simple enough for users to calculate and apply their own t-value if they really want.
- Source sites are now stored as a structured array rather than in a separate SourceSite class (which has been removed).
- The IPython notebooks have been updated accordingly.

@wbinventor Assigning you since there are quite a few changes in the internals of the Python API. You ought to make sure I didn't screw anything up, particularly those things the test suite is missing at the moment.

paulromano commented:
Two more changes I just made:
- 3D voxel plot files are now written to an HDF5 file rather than a raw binary file.
- The 'filetype' dataset in all HDF5 files is now a string. This removes the need to have FILETYPE_* constants in the F90 source files.

paulromano commented:
I realize a few more things are needed before this can be merged:

- [ ] Updates to documentation
- [ ] Updates to scripts that use openmc.statepoint

wbinventor commented:
@paulromano I've only had time to very briefly look over the changes you've made here and I really like the direction your taking with this PR. Switching to HDF5-only really makes things much simpler and opens up all kinds of cool opportunities for downstream data processing. I won't have much time to really dig into this perhaps till middle of next week. 

I do think that we certainly need documentation that at least notes the on-the-fly loading of tally data at the minimum. I also think that some sort of ``Tally.clear_data()`` or ``Tally.flush_results()`` routine to delete the loaded NumPy arrays of tally data would be nice to see. This would allow a user to load/process/delete data from a statepoint file which might exceed the memory capacity on their system. I recognize that this is certainly not a common use case *yet*, but would bring things full circle with the new on-the-fly loading capability. What do you think?

paulromano commented:
A few more things I just updated:
- I added a new IPython notebook demonstrating some basic post processing on tallies and explaining what's going on under the hood. I've editing the "Data Processing and Visualization" section of the documentation to mostly refer to this notebook.
- Removed the openmc-statepoint-histogram script since it only takes a few lines of code with the Python API to replicate and is shown in the notebook I added.
- Added a stub for documentation of the summary file format.
- Added filetype and revision to summary file. S(a,b) table names are now written as an array of strings rather than separate datasets.

Later on, I'd like to take a thorough look at our summary and statepoint formats and decide what we want to change/eliminate. For example, there are many "n_*" datasets that just describe the size of an array. With a structured format like HDF5, those can safely be removed.

paulromano commented:
One last thing -- the "openmc-statepoint-3d" script is probably broken (and probably has been for a while). I'd prefer to leave it for another day and fix it in a separate PR.

paulromano commented:
Some more stuff I just added:
- Renamed `StructuredMesh` derived type to `RegularMesh` to be more precise, since it is not a truly general structured mesh. Mesh type is now written as 'regular' to the statepoint/summary files.
- Ability to use ENDF reaction names for tally scores input, e.g., you can say `<scores>fission (n,2n) (n,p) (n,elastic)`.
- Fixed scoring of elastic scattering, which didn't work before.
- run_mode, filter type, and tally scores written as strings to HDF5 files rather than arbitrary integers.
- The cell surfaces list is written to summary.h5 using user IDs rather than OpenMC internal IDs.
- The openmc.constants module is now gone.
- Various unnecessary datasets removed from summary/statepoint files. Notably, nuclides are not written now to the summary file (not really a summary of the user's input).

I think I'm finally done making changes to this PR!

wbinventor commented:
@paulromano I hope to start looking at starting on Wednesday (sorry for the delay). One concern I have regards the removal of nuclide info from statepoint/summary files. I for one rely heavily on the nuclide info for the tally data processing that I do on a routine basis. Of course, I am prematurely voicing this concern here since I have not yet looked over your code. With your changes, would the nuclide info be reported in the same way in the Pandas DataFrames for tally data as shown in our example IPython Notebooks?

paulromano commented:
The nuclide info should still be reported in the same way, and in fact the statepoint already has that information (i.e. you don't need to call `link_with_summary(...)` to get it) since now nuclide bins are written as an array of strings. Right now, the cross section identifier is reported along with the name of the nuclide (as in "U-235.71c"). Let me know if you think that's ok or if you'd prefer that I remove the identifier.

paulromano commented:
Actually, I just realized it's a little bit problematic to have the cross section identifier there, so I'm going to leave it as it was before, only "U-235". So, the functionality should be exactly the same as before. The only thing you're losing by not having the nuclides in the summary file is that you won't have access to things like `awr` and `kT` which I don't believe were being used anyway. Interrogation of cross section tables is a job better suited for [pyne](https://github.com/pyne/pyne).

wbinventor commented:
Ok, thanks for the explanation, if the nuclide functionality is exactly the same then I'll be happy with it. Look for my comments in a few days.

wbinventor commented on docs/source/pythonapi/examples/post-processing.rst:
```diff
@@ -0,0 +1,13 @@
+
+.. only:: html
+
+    .. notebook:: post-processing.ipynb
```
I'm a fan of your new Notebook, but I'm curious why it is the only one included in this section on post processing?

paulromano replied:
This is really just the name of the notebook (if you have a better non-generic name, feel free to suggest). In the index page for the Python API, you'll see it's listed alongside the pandas dataframes and tally arithmetic notebooks.

wbinventor commented on docs/source/usersguide/output/statepoint.rst:
```diff
@@ -144,14 +142,10 @@ if (run_mode == MODE_EIGENVALUE)
         Unique identifier of the mesh.
 
-    **/tallies/meshes/mesh i/type** (*int*)
+    **/tallies/meshes/mesh i/type** (*char[]*)
```
Is it true that ``i`` here is a monotonically increasing integer starting from 1? Or is it the unique mesh identifier specified by the user (and as indicated by the text above? If if is the user-specified identifier, then the loop here is mis-leading. If if is the internally generated index for the mesh, then that should be indicated for clarity.

paulromano replied:
You're right, it should be the user-defined id. I'll change that.

wbinventor commented on docs/source/usersguide/output/statepoint.rst:
```diff
@@ -182,9 +176,9 @@ if (run_mode == MODE_EIGENVALUE)
 *do i = 1, n_tallies*
 
-    **/tallies/tally i/estimator** (*int*)
+    **/tallies/tally i/estimator** (*char[]*)
```
Ditto, the same comment as before for meshes applies here to tally IDs. I haven't yet made it to the new code in this PR for the statepoint, but was under the impression you had switched to using user-specified IDs in the statepoint file in place of the internally-generated indices.

wbinventor commented on docs/source/usersguide/output/statepoint.rst:
```diff
@@ -182,9 +176,9 @@ if (run_mode == MODE_EIGENVALUE)
+    **/tallies/tally i/estimator** (*char[]*)
 
-        Type of tally estimator: analog (1) or tracklength (2).
+        Type of tally estimator.
```
Why remove the supported tally estimator types?

paulromano replied:
I'll add those back in.

wbinventor commented on docs/source/usersguide/output/statepoint.rst:
```diff
@@ -212,33 +206,31 @@ if (run_mode == MODE_EIGENVALUE)
     **/tallies/tally i/n_score_bins** (*int*)
 
-        Number of scoring bins.
+        Number of scores.
```
I think this was clearer before as "scoring bins" as opposed to "scores." Perhaps it should be expanded to say that some scores may have more than one scoring bins (*e.g.*, "scatter-PN") for clarity.

wbinventor commented on docs/source/usersguide/output/statepoint.rst:
```diff
@@ -212,33 +206,31 @@ if (run_mode == MODE_EIGENVALUE)
 
-            Tallying moment order for Legendre and spherical
-            harmonic tally expansions (*e.g.*, 'P2', 'Y1,2', etc.).
+        Accumulated sum and sum-of-squares for each bin of the i-th tally.
```
Might mention that this is a 2D array and specify how to index into the sum and sum-of-squares.

wbinventor commented on docs/source/usersguide/output/summary.rst:
```diff
@@ -0,0 +1,228 @@
+
+**/date_and_time** (*char[]*)
+
+    Date and time the state point was written.
```
this should be "Date and time the summary was written"

wbinventor commented on docs/source/usersguide/output/summary.rst:
```diff
@@ -0,0 +1,228 @@
+
+do i = 1, n_cells
+
+    **/geometry/cells/cell <uid>/index** (*int*)
```
I assume that each of these elements will be expounded upon in a future PR akin to what is done for the statepoint file?

paulromano replied:
I was originally going to leave it sparse and then fill in details for another PR, but I think it's appropriate to just do it now. I wasn't sure how much of it would change going forward, but after all these changes I have more confidence that the format is starting to stabilize.

wbinventor commented on docs/source/usersguide/output/summary.rst:
```diff
@@ -0,0 +1,228 @@
+
+    **/geometry/cells/cell <uid>/offset** (*int[]*)
+
+    **/geometry/cells/cell <uid>/translated** (*int*)
```
Although I know that "translated" and "rotated" are indeed represented as integers, it might be more transparent to specify them as booleans.

paulromano replied:
Problem is that booleans don't exist in HDF5. I'm guessing this is because C89 had no boolean type. We could write an unsigned short integer, but it's still going to be a zero or one.

paulromano replied:
Rethinking about this, there's no reason we need to write the 'translated' and 'rotated' datasets. Simply just check if 'translation' and 'rotation' are present and if so, read them.

wbinventor commented on docs/source/usersguide/output/summary.rst:
```diff
@@ -0,0 +1,228 @@
+
+**/geometry/n_lattices** (*int*)
+
+do i = 1, n_cells
```
I'm curious why the do loops are included here? This seems like a relic of the old binary files. This might be confusing to new users since the loop index ``i`` is never used in the text within the loop since we are using uids now.

paulromano replied:
Good point. Do you think it's obvious enough that anything with `<uid>` appears multiple times for each unique instance?

wbinventor commented on docs/source/usersguide/output/summary.rst:
```diff
@@ -0,0 +1,228 @@
+
+    **/geometry/lattices/lattice <uid>/universes** (*int[]*)
+
+    if (rectangular lattice)
```
perhaps this should indicated that this references the "type" attribute above?

wbinventor commented on docs/source/usersguide/output/summary.rst:
```diff
@@ -0,0 +1,228 @@
+
+    **/tallies/tally <uid>/nuclides** (*char[][]*)
+
+    **/tallies/tally <uid>/n_score_bins** (*int*)
```
Perhaps you want to leave this documentation sparse for the time being, but I think that some attributes like "total_score_bins" and "n_score_bins" for tallies here would be confusing to any new user.

wbinventor commented on docs/source/usersguide/processing.rst:
```diff
@@ -6,31 +6,33 @@ Data Processing and Visualization
-.. [3] Optional for plotting utilities
+.. [1] Required for most post-processing tasks
+.. [2] Required for reading HDF5 output files
+.. [3] Not used directly by the Python API, but are optional dependencies for a
```
I would say that "pandas" is used directly by the Python API, but is "optional" - minor clarification, you can decide whether or not it is necessary.

wbinventor commented on docs/source/usersguide/processing.rst:
```diff
@@ -183,150 +183,38 @@ doing this will depend on the 3D viewer, but should be straightforward.
-the preferred method of data analysis and visualization.
+many cases the user requires more information about the tally or the run, or has
+to deal with a large number of result values (e.g. for mesh tallies).  In these
+cases, extracting data from the statepoint file via the Python API is the
```
Perhaps link to the Python API reference here

wbinventor commented on docs/source/usersguide/processing.rst:
```diff
@@ -183,150 +183,38 @@ doing this will depend on the 3D viewer, but should be straightforward.
+The :ref:`IPython notebook example <notebook_post_processing>` also demonstrates
+how to plot a mesh tally in two dimensions using the Python API. Note, however,
+that there is also a script distributed with OpenMC, ``openmc-plot-mesh-tally``,
+that interactive GUI to explore and plot mesh tallies for any scores and filter
```
grammatical typo: "that interactive GUI" -> "that provides an interactive GUI"

wbinventor commented on openmc/particle_restart.py:
```diff
@@ -43,28 +39,49 @@ def __init__(self, filename):
+                'filetype'].value.decode() != 'particle restart':
+            raise IOError('{} is not a particle restart file.'.format(filename))
+        if self._f['revision'].value != 1:
+            raise IOError('Particle restart file revision is not consistent.')
```
I think this error message (actually, a lot of our errors messages) would be very opaque to the novice user. Perhaps include the file's revision along with revision for the currently installed version of OpenMC to clarify this message.

paulromano replied:
Very opaque indeed. I'll fix this up.

wbinventor commented on openmc/statepoint.py:
```diff
@@ -94,475 +91,385 @@ def __init__(self, filename):
+                'filetype'].value.decode() != 'statepoint':
+            raise IOError('{} is not a statepoint file.'.format(filename))
+        if self._f['revision'].value != 14:
+            raise IOError('Statepoint revision is not consistent.')
```
Ditto my comment about the revision error message here

wbinventor commented on openmc/statepoint.py:
```diff
@@ -94,475 +91,385 @@ def __init__(self, filename):
+    @property
+    def k_generation(self):
+        if self.run_mode == 'k-eigenvalue':
+            return self._f['k_generation']/value
```
typo: this should be ``.value`` not ``/value``

paulromano replied:
oh wow, yes!

wbinventor commented on openmc/statepoint.py:
```diff
@@ -94,475 +91,385 @@ def __init__(self, filename):
-                # Compute the total number of bins for this Tally
-                num_tot_bins = tally.num_bins
+                # Create Tally object and assign basic properties
+                tally = openmc.Tally(tally_key)
```
Perhaps add the ``id=tally_key`` keyword argument here just to be perfectly clear that the tally keys are the user-specified tally IDs rather than user-specified tally names.

wbinventor replied:
Speaking of which, it doesn't look like the tally label/name is being read here at all. Is that intentional or am I missing something?

paulromano replied:
Hmm, looks like it. I'll have it read the name and pass it upon instantiating.

paulromano replied:
Actually, currently we don't write the name in the statepoint file. Do you think we should add it? I'm leaning excluding metadata like this from the statepoint file that better fits in the summary file.

wbinventor commented on openmc/statepoint.py:
```diff
@@ -94,475 +91,385 @@ def __init__(self, filename):
-    def read_source(self):
-        """Read and store source sites from the statepoint file. By default, source
-        sites are not loaded upon initialization.
+                    if n_bins <= 0:
```
Do you think this error checking is really necessary?

paulromano replied:
I can't imagine it happening, so I'll remove it.

wbinventor commented on openmc/statepoint.py:
```diff
@@ -94,475 +91,385 @@ def __init__(self, filename):
-        """
+                    # Read the bin values
+                    if filter_type in ['energy', 'energyout']:
+                        bins = self._f['{0}{1}/bins'.format(subbase, j)].value
```
It looks like the expressions in each of the ``if-elif-else`` code blocks are the same and can be simplified to a single line of code without need for the conditional.

paulromano replied:
Oh, yay for HDF5!

wbinventor commented on openmc/statepoint.py:
```diff
@@ -94,475 +91,385 @@ def __init__(self, filename):
 
-        """
+                scores = self._f['{0}{1}/score_bins'.format(
+                    base, tally_key)].value
```
I think that adding whitespaces to this line to make it line up with indexing bracket above is easier to read (as you did for ``n_user_scores`` below).

wbinventor commented on openmc/statepoint.py:
```diff
@@ -94,475 +91,385 @@ def __init__(self, filename):
-        for each Tally bin.
+                # Read scattering moment order strings (e.g., P3, Y-1,2, etc.)
+                moments = self._f['{0}{1}/moment_orders'.format(
+                    base, tally_key)].value
```
Add whitespace (or stick to the style guide in other places where there is whitespace)

paulromano replied:
Our style guide says follow PEP8 for Python code. PEP8 says continuation lines should match wrapped elements (like for n_user_scores) or use a hanging indent (like here and for score_bins), in which case the indentation level is appropriate (4 spaces). Really, I'm just justifying my lazy use of emacs auto-indent, which I do believe to implement the correct behavior :)

wbinventor commented on openmc/summary.py:
```diff
@@ -97,23 +67,17 @@ def _read_materials(self):
-
-            sab_names = []
-            sab_xs = []
+            nuclides = self._f['materials'][key]['nuclides'].value
```
Awesome!

wbinventor commented on openmc/summary.py:
```diff
@@ -121,22 +85,21 @@ def _read_materials(self):
+                if 'nat' in name:
+                    nuclide = openmc.Element(name=name, xs=xs)
+                else:
+                    nuclide = openmc.Nuclide(name=name, xs=xs)
```
Rather than use ``nuclide`` for an instance of a ``Nuclide`` or an ``Element``, why not merge the following ``if-else`` statement which adds the ``Nuclide``/``Element`` to the ``Material`` into this ``if-else`` statement.

paulromano replied:
good point

wbinventor commented on openmc/tallies.py:
```diff
@@ -259,24 +264,69 @@ def with_summary(self):
 
     @property
     def sum(self):
+        if not self._statepoint:
```
If you were to simply return the ``sum`` here rather than set it equal to the ``_sum`` class attribute, this would automatically allow for the data to be "cleared" from the tally by the Python garbage collector if and when it is no longer needed (*e.g.*, if the statepoint file is too large to fit in memory, so a user processes tally data in "batches"). This approach would also align well with the approach you used in the statepoint and summary properties.

wbinventor replied:
Actually, this may not be a good idea after all. Tally arithmetic would be difficult (and slow) without attributes dedicated to storing the arrays.

wbinventor commented on openmc/tallies.py:
```diff
@@ -259,24 +264,69 @@ def with_summary(self):
 
     @property
     def sum_sq(self):
+        if not self._statepoint:
```
If you like my recommendation above about just returning the data array, you will need to separately build the ``sum_sq`` array here.

wbinventor commented on openmc/tallies.py:
```diff
@@ -259,24 +264,69 @@ def with_summary(self):
+            self._std_dev = np.zeros_like(self.mean)
+            self._std_dev[nonzero] = np.sqrt((self.sum_sq[nonzero]/n -
+                                              self.mean[nonzero]**2)/(n - 1))
+            self.with_batch_statistics = True
```
If you decide to load tally data on the fly and return it here, the ``with_batch_statistics`` property will no longer be relevant.

wbinventor commented on src/hdf5_interface.F90:
```diff
@@ -1529,11 +1531,94 @@ end subroutine write_string
+    f_ptr = c_loc(buffer(1:1))
+
+    if (using_mpio_device(group_id)) then
+#ifdef PHDF5
```
If I understand correctly, if using MPI but not using PHDF5 this ``if`` will enclose an empty block and the string will not be read. Is that intentional?

paulromano replied:
Not quite. If you're not using PHDF5, when the file is opened, the file driver will just be the default one which uses POSIX file-system functions rather than the MPIO driver which relies on MPI-IO for accessing files. So, if only MPI, `using_mpio_devce(group_id)` will be false.

wbinventor commented on src/hdf5_interface.F90:
```diff
@@ -1551,33 +1637,122 @@ subroutine read_string(group_id, name, buffer, indep)
+    f_ptr = c_loc(buffer(1)(1:1))
+
+    if (using_mpio_device(group_id)) then
+#ifdef PHDF5
```
Ditto the empty ``if`` coupled with the ``ifdef`` macro seems to be an issue.

wbinventor commented on src/hdf5_interface.F90:
```diff
@@ -1551,33 +1637,122 @@ subroutine read_string(group_id, name, buffer, indep)
       call h5pset_dxpl_mpio_f(plist, data_xfer_mode, hdf5_err)
-      call h5dread_vl_f(dset, H5T_STRING, str_tmp, dims2, dims1, hdf5_err, &
-           mem_space_id=dspace, xfer_prp=plist)
+      call h5dread_f(dset, memtype, f_ptr, hdf5_err, mem_space_id=dspace, &
```
Ditto here and elsewhere, I will stop commenting since there may be a reason for this.

wbinventor commented on src/particle_restart_write.F90:
```diff
@@ -40,13 +40,22 @@ subroutine write_particle_restart(p)
+    case (MODE_EIGENVALUE)
+      call write_dataset(file_id, 'run_mode', 'k-eigenvalue')
+    case (MODE_PLOTTING)
+      call write_dataset(file_id, 'run_mode', 'plot')
```
Would one every do run a particle restart in plotting mode?

paulromano replied:
Well, plotting does use a fake particle to find cells, but I think it's safe to say no, plotting mode should never produce a particle restart file.

wbinventor commented on src/state_point.F90:
```diff
@@ -40,19 +41,20 @@ module state_point
+    integer(HID_T) :: filter_group
+    character(20), allocatable :: str_array(:)
+    character(MAX_FILE_LEN)    :: filename
+    type(RegularMesh), pointer :: meshp
```
just curious why this is ``meshp`` rather than ``mesh``?

paulromano replied:
The PGI compiler has some weird bug where if you have a local variable that is the same name as a module, it sometimes produces an error... yes, only sometimes, and this was one of those times.

wbinventor commented on src/state_point.F90:
```diff
@@ -90,7 +92,16 @@ subroutine write_state_point()
+        call write_dataset(file_id, "run_mode", "fixed source")
+      case (MODE_EIGENVALUE)
+        call write_dataset(file_id, "run_mode", "k-eigenvalue")
+      case (MODE_PLOTTING)
```
Can a statepoint even be created in plotting mode?

wbinventor commented on src/state_point.F90:
```diff
@@ -170,8 +181,7 @@ subroutine write_state_point()
           call write_dataset(mesh_group, "id", meshp%id)
-          call write_dataset(mesh_group, "type", meshp%type)
-          call write_dataset(mesh_group, "n_dimension", meshp%n_dimension)
+          call write_dataset(mesh_group, "type", "regular")
```
Why not keep as the more general ``mesh%type`` assuming we will one day have hex and/or other structured tally mesh types?

wbinventor commented on src/state_point.F90:
```diff
@@ -223,7 +240,28 @@ subroutine write_state_point()
 
-            call write_dataset(filter_group, "type", tally%filters(j)%type)
+            ! Write name of type
+            select case (tally%filters(j)%type)
```
I'm not sure if it would simplify the code elsewhere, but it seems that the use of a hash table / dictionary to store the filter enumerated types mapped to string types would be cleaner than a ``case`` statement which will inevitably grow larger in the future.

paulromano replied:
That is very true. I may do a branch in the near future cleaning up things like this.

wbinventor commented on src/state_point.F90:
```diff
@@ -238,59 +276,108 @@ subroutine write_state_point()
-              key_array(j) = nuclides(tally%nuclide_bins(j))%zaid
+              i_list = nuclides(tally%nuclide_bins(j))%listing
+              i_xs = index(xs_listings(i_list)%alias, '.')
+              if (i_xs > 0) then
```
It might be nice to include a comment around this ``if-else`` to clarify what ``i_xs`` is doing here.

wbinventor commented on src/state_point.F90:
```diff
@@ -238,59 +276,108 @@ subroutine write_state_point()
           call write_dataset(tally_group, "n_score_bins", tally%n_score_bins)
-          call write_dataset(tally_group, "score_bins", tally%score_bins)
+          allocate(str_array(size(tally%score_bins)))
+          do j = 1, size(tally%score_bins)
```
Again, I like the idea of a dictionary mapping score enumerated types to string definitions, but it is up to you.

wbinventor commented on src/state_point.F90:
```diff
@@ -866,14 +790,12 @@ subroutine load_state_point()
 
           ! Read sum and sum_sq for each bin
           tally_group = open_group(tallies_group, "tally " // &
-               trim(to_str(curr_key)))
```
How does this error out if the user adds a new tally to their "tallies.xml" file? Or in the reverse scenario if a tally has been removed from the "tallies.xml" file, no error will be reported and this tally will simply not be reported in later statepoints. What do you think about some sort of error checking capability here to deal with this?

wbinventor replied:
Of course it is in reality much more complicated than this, if a user changes a tally in "tallies.xml" (*e.g.*, removes/adds a score) and tries to load from a statepoint no error message is reported.

paulromano replied:
There are probably many things that could go wrong in the hands of a user who changes their input and then tries to restart. I think we need should decide on an overall strategy for what we allow to be changed in a restart run before trying to catch errors.

wbinventor commented on src/summary.F90:
```diff
@@ -537,46 +528,100 @@ subroutine write_tallies(file_id)
         select case (t%filters(j)%type)
         case(FILTER_UNIVERSE)
-          call write_dataset(filter_group, "type_name", "universe")
+          call write_dataset(filter_group, "type", "universe")
```
If you choose to use a dictionary to store filter types, remember to reuse it here.

wbinventor commented on src/summary.F90:
```diff
@@ -537,46 +528,100 @@ subroutine write_tallies(file_id)
+      allocate(str_array(size(t%score_bins)))
+      do j = 1, size(t%score_bins)
+        select case(t%score_bins(j))
+        case (SCORE_FLUX)
```
Ditto reuse score type dictionary here.

wbinventor commented on src/input_xml.F90:
```diff
@@ -2188,9 +2188,11 @@ subroutine read_tallies_xml()
-        m % type = LATTICE_RECT
-      case ('hex', 'hexagon', 'hexagonal')
-        m % type = LATTICE_HEX
+        call warning("Mesh type '" // trim(temp_str) // "' is deprecated. &
```
Several of the example inputs still include "rectangular" meshes.

wbinventor replied:
In addition, the "python/lattice/nested", "python/lattice/simple" and "python/pincell" examples fail since the Python API no longer accepts the "regular" mesh option.

wbinventor commented:
@paulromano I've fully reviewed this PR and have tested the changes with the example inputs and IPython Notebooks. The changes you've made here and in recent PRs have greatly streamlined the I/O in the code. I wish OpenMC had been HDF5-only a year ago, it would have shaved a 1-2 months off of my PhD :-) This PR is ready to go once you've addressed my comments, unless any other developers want to chime in first.

paulromano commented:
Thanks for the thorough review @wbinventor! I think I've addressed all the comments that need addressing. Take a look again and let me know if there's anything left that I missed.

paulromano commented:
@wbinventor I've addressed your second round of comments.
